[fusion] add composable fusion pass framework

DevashishLal-CB opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

<!-- Describe the purpose and goals of this pull request. -->
Initial implementation of the changes proposed in #10118 

## Modifications

<!-- Detail the changes made in this pull request. -->
This PR adds the fusion passes and integration tests for them

Passes added
- gate_up project + silu and mul (triton kernel, [cutlass kernel](https://github.com/NVIDIA/cutlass/tree/main/examples/45_dual_gemm) is possible)
- gate_up project + silu and mul + quant (triton kernel, [cutlass kernel](https://github.com/NVIDIA/cutlass/tree/main/examples/45_dual_gemm) is possible)
- rmsnorm + quant (kernel from vllm)
- fused_add_rmsnorm + quant (kernel from vllm)

For fusion passes to work with cuda graph runner I had to get rid for the model patching (or I could rewrite the pass with the pattern functions looking for pure pytorch code, we should avoid this model patching as it will interfere with the compilation process)

I have also added `model_bench.py`, the idea with this is to provide a stripped down sglang runtime where each layer can be instantiated in isolation helping write integration and accuracy tests from fusion passes and fused kernels

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

For llama 3.1 8B FP8, BS1, ISL 1024, OSL 1024. 6.2% Gains

```bash
# Dual GEMM Quant + RMSNorm Quant
Benchmark ...
Prefill. latency: 0.01871 s, throughput:  54740.79 token/s
Decode 0. Batch size: 1, latency: 0.00531 s, throughput:    188.24 token/s
Decode 1. Batch size: 1, latency: 0.00535 s, throughput:    186.91 token/s
Decode 2. Batch size: 1, latency: 0.00536 s, throughput:    186.62 token/s
Decode 3. Batch size: 1, latency: 0.00533 s, throughput:    187.63 token/s
Decode 4. Batch size: 1, latency: 0.00529 s, throughput:    189.06 token/s
Decode.  median latency: 0.00529 s, median throughput:    189.14 token/s
Total. latency:  5.445 s, throughput:    376.13 token/s

# RMSNorm Quant
Benchmark ...
Prefill. latency: 0.01944 s, throughput:  52668.28 token/s
Decode 0. Batch size: 1, latency: 0.00552 s, throughput:    181.00 token/s
Decode 1. Batch size: 1, latency: 0.00543 s, throughput:    184.12 token/s
Decode 2. Batch size: 1, latency: 0.00545 s, throughput:    183.35 token/s
Decode 3. Batch size: 1, latency: 0.00539 s, throughput:    185.65 token/s
Decode 4. Batch size: 1, latency: 0.00544 s, throughput:    183.94 token/s
Decode.  median latency: 0.00539 s, median throughput:    185.64 token/s
Total. latency:  5.546 s, throughput:    369.31 token/s

# Default Torch Compile + Graph Break Fixes
Benchmark ...
Prefill. latency: 0.01865 s, throughput:  54917.17 token/s
Decode 0. Batch size: 1, latency: 0.00566 s, throughput:    176.68 token/s
Decode 1. Batch size: 1, latency: 0.00562 s, throughput:    178.03 token/s
Decode 2. Batch size: 1, latency: 0.00560 s, throughput:    178.63 token/s
Decode 3. Batch size: 1, latency: 0.00559 s, throughput:    178.76 token/s
Decode 4. Batch size: 1, latency: 0.00561 s, throughput:    178.39 token/s
Decode.  median latency: 0.00561 s, median throughput:    178.13 token/s
Total. latency:  5.765 s, throughput:    355.25 token/s
```

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Logs
MM + Silu and Mul fusion
```shell
 ===== Before_FusedActivationPass =====
 <eval_with_key>.198 from /usr/local/lib/python3.12/dist-packages/torch/fx/experimental/proxy_tensor.py:1301 in wrapped class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "bf16[1, 2048][2048, 1]cuda:0", arg1_1: "bf16[1, 2048][2048, 1]cuda:0", arg2_1: "bf16[2048][1]cuda:0", arg3_1: "bf16[16384, 2048][2048, 1]cuda:0", arg4_1: "bf16[2048, 8192][8192, 1]cuda:0"):
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:81 in fused_add_rmsnorm, code: torch.ops.sgl_kernel.fused_add_rmsnorm.default(
        auto_functionalized_v2 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.fused_add_rmsnorm.default, weight = arg2_1, eps = 1e-05, enable_pdl = False, _input_base_index = 0, _residual_base_index = 1, _all_bases = [arg0_1, arg1_1]);  arg2_1 = None
        getitem_1: "bf16[1, 2048][2048, 1]cuda:0" = auto_functionalized_v2[1]
        getitem_2: "bf16[1, 2048][2048, 1]cuda:0" = auto_functionalized_v2[2];  auto_functionalized_v2 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/unquant.py:135 in apply, code: return F.linear(x, layer.weight, bias)
        permute: "bf16[2048, 16384][1, 2048]cuda:0" = torch.ops.aten.permute.default(arg3_1, [1, 0]);  arg3_1 = None
        mm: "bf16[1, 16384][16384, 1]cuda:0" = torch.ops.aten.mm.default(getitem_1, permute);  permute = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/activation.py:69 in forward_cuda, code: out = torch.empty(output_shape, dtype=x.dtype, device=x.device)
        empty: "bf16[1, 8192][8192, 1]cuda:0" = torch.ops.aten.empty.memory_format([1, 8192], dtype = torch.bfloat16, device = device(type='cuda', index=0), pin_memory = False)
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:183 in silu_and_mul, code: torch.ops.sgl_kernel.silu_and_mul.default(out, input)
        auto_functionalized_v2_1 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.silu_and_mul.default, input = mm, _out_base_index = 0, _all_bases = [empty]);  mm = empty = None
        getitem_4: "bf16[1, 8192][8192, 1]cuda:0" = auto_functionalized_v2_1[1];  auto_functionalized_v2_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/unquant.py:135 in apply, code: return F.linear(x, layer.weight, bias)
        permute_1: "bf16[8192, 2048][1, 8192]cuda:0" = torch.ops.aten.permute.default(arg4_1, [1, 0]);  arg4_1 = None
        mm_1: "bf16[1, 2048][2048, 1]cuda:0" = torch.ops.aten.mm.default(getitem_4, permute_1);  getitem_4 = permute_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:81 in fused_add_rmsnorm, code: torch.ops.sgl_kernel.fused_add_rmsnorm.default(
        copy_: "bf16[1, 2048][2048, 1]cuda:0" = torch.ops.aten.copy_.default(arg0_1, getitem_1);  arg0_1 = getitem_1 = copy_ = None
        copy__1: "bf16[1, 2048][2048, 1]cuda:0" = torch.ops.aten.copy_.default(arg1_1, getitem_2);  arg1_1 = getitem_2 = copy__1 = None
        return (mm_1,)
        
[2025-09-17 02:10:27] TRACED GRAPH
 ===== After_FusedActivationPass =====
 <eval_with_key>.198 from /usr/local/lib/python3.12/dist-packages/torch/fx/experimental/proxy_tensor.py:1301 in wrapped class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "bf16[1, 2048][2048, 1]cuda:0", arg1_1: "bf16[1, 2048][2048, 1]cuda:0", arg2_1: "bf16[2048][1]cuda:0", arg3_1: "bf16[16384, 2048][2048, 1]cuda:0", arg4_1: "bf16[2048, 8192][8192, 1]cuda:0"):
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:81 in fused_add_rmsnorm, code: torch.ops.sgl_kernel.fused_add_rmsnorm.default(
        auto_functionalized_v2 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.fused_add_rmsnorm.default, weight = arg2_1, eps = 1e-05, enable_pdl = False, _input_base_index = 0, _residual_base_index = 1, _all_bases = [arg0_1, arg1_1]);  arg2_1 = None
        getitem_1: "bf16[1, 2048][2048, 1]cuda:0" = auto_functionalized_v2[1]
        getitem_2: "bf16[1, 2048][2048, 1]cuda:0" = auto_functionalized_v2[2];  auto_functionalized_v2 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/unquant.py:135 in apply, code: return F.linear(x, layer.weight, bias)
        permute: "bf16[2048, 16384][1, 2048]cuda:0" = torch.ops.aten.permute.default(arg3_1, [1, 0]);  arg3_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/activation.py:69 in forward_cuda, code: out = torch.empty(output_shape, dtype=x.dtype, device=x.device)
        empty: "bf16[1, 8192][8192, 1]cuda:0" = torch.ops.aten.empty.memory_format([1, 8192], dtype = torch.bfloat16, device = device(type='cuda', index=0), pin_memory = False);  empty = None
        
        # No stacktrace found for following nodes
        fused_swiglu_default: "bf16[1, 8192][8192, 1]cuda:0" = torch.ops.sglang.fused_swiglu.default(getitem_1, permute);  permute = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/unquant.py:135 in apply, code: return F.linear(x, layer.weight, bias)
        permute_1: "bf16[8192, 2048][1, 8192]cuda:0" = torch.ops.aten.permute.default(arg4_1, [1, 0]);  arg4_1 = None
        mm_1: "bf16[1, 2048][2048, 1]cuda:0" = torch.ops.aten.mm.default(fused_swiglu_default, permute_1);  fused_swiglu_default = permute_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:81 in fused_add_rmsnorm, code: torch.ops.sgl_kernel.fused_add_rmsnorm.default(
        copy_: "bf16[1, 2048][2048, 1]cuda:0" = torch.ops.aten.copy_.default(arg0_1, getitem_1);  arg0_1 = getitem_1 = copy_ = None
        copy__1: "bf16[1, 2048][2048, 1]cuda:0" = torch.ops.aten.copy_.default(arg1_1, getitem_2);  arg1_1 = getitem_2 = copy__1 = None
        return (mm_1,)
        
[2025-09-17 02:10:27] FusedActivationPass completed in 19.8 ms, matched 1 times
```

MM + Silu and Mul + Quant (I have a small diff to use sgl_per_tensor_quant_fp8 for quant instead of the triton quant kernel, will add support for the default quant kernel before merge)
```shell
[2025-09-17 02:10:39] TRACED GRAPH
 ===== Before_FusedActivationPass =====
 <eval_with_key>.56 from /usr/local/lib/python3.12/dist-packages/torch/fx/experimental/proxy_tensor.py:1301 in wrapped class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f16[6, 4096][4096, 1]cuda:0", arg1_1: "f16[6, 4096][4096, 1]cuda:0", arg2_1: "f16[4096][1]cuda:0", arg3_1: "f8e4m3fn[4096, 22016][1, 4096]cuda:0", arg4_1: "f32[][]cuda:0", arg5_1: "f32[][]cuda:0", arg6_1: "f8e4m3fn[11008, 4096][1, 11008]cuda:0", arg7_1: "f32[][]cuda:0", arg8_1: "f32[][]cuda:0"):
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:81 in fused_add_rmsnorm, code: torch.ops.sgl_kernel.fused_add_rmsnorm.default(
        auto_functionalized_v2 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.fused_add_rmsnorm.default, weight = arg2_1, eps = 1e-05, enable_pdl = False, _input_base_index = 0, _residual_base_index = 1, _all_bases = [arg0_1, arg1_1]);  arg2_1 = None
        getitem_1: "f16[6, 4096][4096, 1]cuda:0" = auto_functionalized_v2[1]
        getitem_2: "f16[6, 4096][4096, 1]cuda:0" = auto_functionalized_v2[2];  auto_functionalized_v2 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/fp8_kernel.py:1417 in scaled_fp8_quant, code: output = torch.empty(shape, device=input.device, dtype=fp8_dtype)
        empty: "f8e4m3fn[6, 4096][4096, 1]cuda:0" = torch.ops.aten.empty.memory_format([6, 4096], dtype = torch.float8_e4m3fn, device = device(type='cuda', index=0), pin_memory = False)
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/gemm.py:136 in sgl_per_tensor_quant_fp8, code: torch.ops.sgl_kernel.sgl_per_tensor_quant_fp8.default(
        auto_functionalized_v2_1 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.sgl_per_tensor_quant_fp8.default, input = getitem_1, output_s = arg5_1, is_static = True, _output_q_base_index = 0, _all_bases = [empty]);  empty = None
        getitem_4: "f8e4m3fn[6, 4096][4096, 1]cuda:0" = auto_functionalized_v2_1[1];  auto_functionalized_v2_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/fp8_utils.py:803 in apply_fp8_linear, code: output = torch._scaled_mm(
        _scaled_mm: "f16[6, 22016][22016, 1]cuda:0" = torch.ops.aten._scaled_mm.default(getitem_4, arg3_1, arg5_1, arg4_1, None, None, torch.float16);  getitem_4 = arg3_1 = arg5_1 = arg4_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/activation.py:69 in forward_cuda, code: out = torch.empty(output_shape, dtype=x.dtype, device=x.device)
        empty_1: "f16[6, 11008][11008, 1]cuda:0" = torch.ops.aten.empty.memory_format([6, 11008], dtype = torch.float16, device = device(type='cuda', index=0), pin_memory = False)
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:183 in silu_and_mul, code: torch.ops.sgl_kernel.silu_and_mul.default(out, input)
        auto_functionalized_v2_2 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.silu_and_mul.default, input = _scaled_mm, _out_base_index = 0, _all_bases = [empty_1]);  _scaled_mm = empty_1 = None
        getitem_6: "f16[6, 11008][11008, 1]cuda:0" = auto_functionalized_v2_2[1];  auto_functionalized_v2_2 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/fp8_kernel.py:1417 in scaled_fp8_quant, code: output = torch.empty(shape, device=input.device, dtype=fp8_dtype)
        empty_2: "f8e4m3fn[6, 11008][11008, 1]cuda:0" = torch.ops.aten.empty.memory_format([6, 11008], dtype = torch.float8_e4m3fn, device = device(type='cuda', index=0), pin_memory = False)
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/gemm.py:136 in sgl_per_tensor_quant_fp8, code: torch.ops.sgl_kernel.sgl_per_tensor_quant_fp8.default(
        auto_functionalized_v2_3 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.sgl_per_tensor_quant_fp8.default, input = getitem_6, output_s = arg8_1, is_static = True, _output_q_base_index = 0, _all_bases = [empty_2]);  getitem_6 = empty_2 = None
        getitem_8: "f8e4m3fn[6, 11008][11008, 1]cuda:0" = auto_functionalized_v2_3[1];  auto_functionalized_v2_3 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/fp8_utils.py:803 in apply_fp8_linear, code: output = torch._scaled_mm(
        _scaled_mm_1: "f16[6, 4096][4096, 1]cuda:0" = torch.ops.aten._scaled_mm.default(getitem_8, arg6_1, arg8_1, arg7_1, None, None, torch.float16);  getitem_8 = arg6_1 = arg8_1 = arg7_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:81 in fused_add_rmsnorm, code: torch.ops.sgl_kernel.fused_add_rmsnorm.default(
        copy_: "f16[6, 4096][4096, 1]cuda:0" = torch.ops.aten.copy_.default(arg0_1, getitem_1);  arg0_1 = getitem_1 = copy_ = None
        copy__1: "f16[6, 4096][4096, 1]cuda:0" = torch.ops.aten.copy_.default(arg1_1, getitem_2);  arg1_1 = getitem_2 = copy__1 = None
        return (_scaled_mm_1,)
        
[2025-09-17 02:10:39] TRACED GRAPH
 ===== After_FusedActivationPass =====
 <eval_with_key>.56 from /usr/local/lib/python3.12/dist-packages/torch/fx/experimental/proxy_tensor.py:1301 in wrapped class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f16[6, 4096][4096, 1]cuda:0", arg1_1: "f16[6, 4096][4096, 1]cuda:0", arg2_1: "f16[4096][1]cuda:0", arg3_1: "f8e4m3fn[4096, 22016][1, 4096]cuda:0", arg4_1: "f32[][]cuda:0", arg5_1: "f32[][]cuda:0", arg6_1: "f8e4m3fn[11008, 4096][1, 11008]cuda:0", arg7_1: "f32[][]cuda:0", arg8_1: "f32[][]cuda:0"):
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:81 in fused_add_rmsnorm, code: torch.ops.sgl_kernel.fused_add_rmsnorm.default(
        auto_functionalized_v2 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.fused_add_rmsnorm.default, weight = arg2_1, eps = 1e-05, enable_pdl = False, _input_base_index = 0, _residual_base_index = 1, _all_bases = [arg0_1, arg1_1]);  arg2_1 = None
        getitem_1: "f16[6, 4096][4096, 1]cuda:0" = auto_functionalized_v2[1]
        getitem_2: "f16[6, 4096][4096, 1]cuda:0" = auto_functionalized_v2[2];  auto_functionalized_v2 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/fp8_kernel.py:1417 in scaled_fp8_quant, code: output = torch.empty(shape, device=input.device, dtype=fp8_dtype)
        empty: "f8e4m3fn[6, 4096][4096, 1]cuda:0" = torch.ops.aten.empty.memory_format([6, 4096], dtype = torch.float8_e4m3fn, device = device(type='cuda', index=0), pin_memory = False)
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/gemm.py:136 in sgl_per_tensor_quant_fp8, code: torch.ops.sgl_kernel.sgl_per_tensor_quant_fp8.default(
        auto_functionalized_v2_1 = torch.ops.higher_order.auto_functionalized_v2(torch.ops.sgl_kernel.sgl_per_tensor_quant_fp8.default, input = getitem_1, output_s = arg5_1, is_static = True, _output_q_base_index = 0, _all_bases = [empty]);  empty = None
        getitem_4: "f8e4m3fn[6, 4096][4096, 1]cuda:0" = auto_functionalized_v2_1[1];  auto_functionalized_v2_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/activation.py:69 in forward_cuda, code: out = torch.empty(output_shape, dtype=x.dtype, device=x.device)
        empty_1: "f16[6, 11008][11008, 1]cuda:0" = torch.ops.aten.empty.memory_format([6, 11008], dtype = torch.float16, device = device(type='cuda', index=0), pin_memory = False);  empty_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/fp8_kernel.py:1417 in scaled_fp8_quant, code: output = torch.empty(shape, device=input.device, dtype=fp8_dtype)
        empty_2: "f8e4m3fn[6, 11008][11008, 1]cuda:0" = torch.ops.aten.empty.memory_format([6, 11008], dtype = torch.float8_e4m3fn, device = device(type='cuda', index=0), pin_memory = False);  empty_2 = None
        
        # No stacktrace found for following nodes
        fused_swiglu_default: "f8e4m3fn[6, 11008][11008, 1]cuda:0" = torch.ops.sglang.fused_swiglu.default(getitem_4, arg3_1, arg5_1, arg4_1, arg8_1);  getitem_4 = arg3_1 = arg5_1 = arg4_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sglang/srt/layers/quantization/fp8_utils.py:803 in apply_fp8_linear, code: output = torch._scaled_mm(
        _scaled_mm_1: "f16[6, 4096][4096, 1]cuda:0" = torch.ops.aten._scaled_mm.default(fused_swiglu_default, arg6_1, arg8_1, arg7_1, None, None, torch.float16);  fused_swiglu_default = arg6_1 = arg8_1 = arg7_1 = None
        
         # File: /usr/local/lib/python3.12/dist-packages/sgl_kernel/elementwise.py:81 in fused_add_rmsnorm, code: torch.ops.sgl_kernel.fused_add_rmsnorm.default(
        copy_: "f16[6, 4096][4096, 1]cuda:0" = torch.ops.aten.copy_.default(arg0_1, getitem_1);  arg0_1 = getitem_1 = copy_ = None
        copy__1: "f16[6, 4096][4096, 1]cuda:0" = torch.ops.aten.copy_.default(arg1_1, getitem_2);  arg1_1 = getitem_2 = copy__1 = None
        return (_scaled_mm_1,)
        
[2025-09-17 02:10:39] FusedActivationPass completed in 17.4 ms, matched 1 times
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).













<!-- pr-states:start -->
---
### CI States

Latest PR Test (Base): <!-- slot:pr-test:start -->:x: **Missing `run-ci` label** -- add it to run CI tests.<!-- slot:pr-test:end -->
Latest PR Test (Extra): <!-- slot:pr-test-extra:start -->:x: **Blocked** -- `run-ci` is required first.<!-- slot:pr-test-extra:end -->
<!-- pr-states:end -->

DevashishLal-CB commented:
Things Pending as of now
- [ ] Cuda graph support, currently the fused activation pass doesn't work with cuda graph enabled
- [ ] RMSNorm Quant fusion pass
- [ ] Unit tests for fusion passes

DevashishLal-CB committed:
[fusion][1/N] add fusion pass manager and base fusion pass

Add initial implementation of a configurable fusion pass
manager as well as basic building blocks for creating
new fusion passes

rfc: https://github.com/sgl-project/sglang/issues/10118

Signed-off-by: Devashish Lal <devashish@rivosinc.com>

DevashishLal-CB committed:
[fusion][2/N] add fused activation pass with triton kernel

fused activation pass fuses the mm + activation (silu_and_mul) + quant
into a single triton kernel, used in MLP layers of the models

adapts fused swiglu from https://arxiv.org/abs/2002.05202

added model factory to intialize individual layers for tests

rfc: https://github.com/sgl-project/sglang/issues/10118

Signed-off-by: Devashish Lal <devashish@rivosinc.com>

DevashishLal-CB committed:
[fusion][3/N] add rmsnorm and quant fusion pass

rmsnorm and fused_add rmsnorm can be fused with the
quantization following them using the quantized rmsnorm
kernel from vllm, hence this pass depends on vllm for now

rfc: https://github.com/sgl-project/sglang/issues/10118

Signed-off-by: Devashish Lal <devashish@rivosinc.com>

BBuf commented:
Can we add a sgl-kernel fuse kernel pass example? Such as [`topk_softmax`](https://github.com/sgl-project/sglang/blob/main/sgl-kernel/python/sgl_kernel/moe.py#L28)

DevashishLal-CB committed:
[fusion][4/N] added topk softmax fusion example and cleanup

test_fusion_example shows a how to write a fusion pass along
with it't test in isolation

rfc: https://github.com/sgl-project/sglang/issues/10118

Signed-off-by: Devashish Lal <devashish@rivosinc.com>

BBuf commented:
> > Can we add a sgl-kernel fuse kernel pass example? Such as [`topk_softmax`](https://github.com/sgl-project/sglang/blob/main/sgl-kernel/python/sgl_kernel/moe.py#L28)
> 
> @BBuf Added the example for topk_softmax fusion, Also added rmsnorm_quant fusion pass with tests
> 
> This MR is ready for review, will look into cuda graph support and do it as a separate MR
> 
> Will collaborate with @yuan-luo

Cool, we'll review ASAP.

yuan-luo commented:
https://github.com/sgl-project/sglang/pull/10987

yuan-luo commented on python/sglang/srt/compilation/fusion/fusion_config.py:
```diff
@@ -0,0 +1,76 @@
+# limitations under the License.
+# ==============================================================================
+
+import logging
```
We'd better put this configuration file in the python/sglang/srt/configs/ directory.

DevashishLal-CB committed:
[fusion][5/N] uprev and integration with piecewise cuda graphs

rebased with origin/main and integrated with piecewise cuda graph
changes, fusions and piecewise graphs work independently but with
both enabled fusions only work for decode

other changes included
- rmsnorm kernels adopted from vllm
- updates to model bench used for integration testing of fusions
- cuda graph compatibility for fusions passes

rfc: https://github.com/sgl-project/sglang/issues/10118
Signed-off-by: Devashish Lal <devashish@rivosinc.com>

BLaZeKiLL committed:
Merge branch 'main' into gh/dlal/sgl-fusion

BLaZeKiLL committed:
Merge branch 'main' into gh/dlal/sgl-fusion

BLaZeKiLL committed:
Merge branch 'main' into gh/dlal/sgl-fusion

BLaZeKiLL committed:
[fusion][6/N] uprev and fix fusion passes

Signed-off-by: Devashish Lal <laldevashish@gmail.com>

BLaZeKiLL committed:
[fusion][7/N] add support for flashinfer rmsnorm + quant fused kernels

Signed-off-by: Devashish Lal <laldevashish@gmail.com>

BLaZeKiLL commented:
Some performance numbers from sglang on an RTX 5090 running llama 3.1 8b fp8 on a 16 prompt benchmark for the rmsnorm_quant fusion pass

| server config | throughput (tok/sec |
|--------|--------|
| cuda graph | 503.25 |
| cuda graph + torch compile | 600.37 |
| cuda graph + torch compile + fusion (vllm kernels) | 612.58 |
| cuda graph + torch compile + fusion (flashinfer kernels) | 619.11 |

Commit:
Merge branch 'main' into gh/dlal/sgl-fusion

Commit:
Merge remote-tracking branch 'upstream/main' into gh/dlal/sgl-fusion

Commit:
[fusion][8/N] Add composable fusion pass framework

Signed-off-by: Devashish Lal <devcode@fb.com>

Commit:
Merge remote-tracking branch 'upstream/main' into gh/dlal/sgl-fusion

Commit:
[fusion][9/N] add norm quant jit kenrels and benchmarks

Signed-off-by: Devashish Lal <devcode@fb.com>

Commit:
Merge remote-tracking branch 'upstream/main' into gh/dlal/sgl-fusion

Commit:
[fusion][9/N] rmsnorm_quant, fused_add_rmsnorm_quant jit kernels

these kernels are faster for all benchmarks when compared
against aot sglang, fused flashinfer (cutedsl) and unfused impl

Signed-off-by: Devashish Lal <devcode@fb.com>

Commit:
[fusion][10/N] Remove graph breaks and ensure full graph compilation

Signed-off-by: Devashish Lal <devcode@fb.com>

Commit:
[fusion][11/N] Add cutedsl dual gemm kernel

Signed-off-by: Devashish Lal <devcode@fb.com>

Commit:
Merge remote-tracking branch 'upstream/main' into gh/dlal/sgl-fusion

Commit:
[fusion][12/N] Uprev and Cleanup

Signed-off-by: Devashish Lal <devcode@fb.com>

Commit:
Merge remote-tracking branch 'upstream/main' into gh/dlal/sgl-fusion

Commit:
[fusion][13/N] Uprev and Cleanup

Signed-off-by: Devashish Lal <devcode@fb.com>

Commit:
[fusion][14/N] code clean up

Signed-off-by: Devashish Lal <devcode@fb.com>

Commit:
stash

Signed-off-by: Devashish Lal <devcode@fb.com>
